external/esp_idf_port/esp32/spi_ram/spiram_psram.c

```diff
@@ -0,0 +1,721 @@
+        psram_set_qio_write_mode(spi_num);
+        psram_set_qio_read_mode(spi_num);
+    }
+
+    //Wait for SPI0 to idle
+    while ( READ_PERI_REG(SPI_EXT2_REG(0)) != 0);
+    DPORT_SET_PERI_REG_MASK(DPORT_HOST_INF_SEL_REG, 1 << 14);
+
+    // Start send data
+    SET_PERI_REG_MASK(SPI_CMD_REG(spi_num), SPI_USR);
+    while ((READ_PERI_REG(SPI_CMD_REG(spi_num)) & SPI_USR));
+    DPORT_CLEAR_PERI_REG_MASK(DPORT_HOST_INF_SEL_REG, 1 << 14);
+
+    //recover spi mode
+    SET_PERI_REG_BITS(SPI_USER_REG(spi_num), (pRxData ? SPI_FWRITE_DUAL_M : 0xf), mode_backup, SPI_FWRITE_DUAL_S);
+    CLEAR_PERI_REG_MASK(SPI_CTRL_REG(spi_num), (SPI_FREAD_DIO_M | SPI_FREAD_DUAL_M | SPI_FREAD_QUAD_M | SPI_FREAD_QIO_M));
+    SET_PERI_REG_MASK(SPI_CTRL_REG(spi_num), rd_mode_backup);
+
+    //return cs to cs0
+    SET_PERI_REG_MASK(SPI_PIN_REG(PSRAM_SPI_1), SPI_CS1_DIS_M);
+    CLEAR_PERI_REG_MASK(SPI_PIN_REG(PSRAM_SPI_1), SPI_CS0_DIS_M);
+
+    if (pRxData) {
+        int idx = 0;
+        // Read data out
+        do {
+            *pRxData++ = READ_PERI_REG(SPI_W0_REG(spi_num) + (idx << 2));
+        } while (++idx < ((rxByteLen / 4) + ((rxByteLen % 4) ? 1 : 0)));
+    }
+}
+
+static uint32_t backup_usr[3];
+static uint32_t backup_usr1[3];
+static uint32_t backup_usr2[3];
+
+
+
+//setup spi command/addr/data/dummy in user mode
+static int psram_cmd_config(psram_spi_num_t spi_num, psram_cmd_t* pInData)
+{
+    while (READ_PERI_REG(SPI_CMD_REG(spi_num)) & SPI_USR);
+    backup_usr[spi_num]=READ_PERI_REG(SPI_USER_REG(spi_num));
+    backup_usr1[spi_num]=READ_PERI_REG(SPI_USER1_REG(spi_num));
+    backup_usr2[spi_num]=READ_PERI_REG(SPI_USER2_REG(spi_num));
+    // Set command by user.
+    if (pInData->cmdBitLen != 0) {
+        // Max command length 16 bits.
+        SET_PERI_REG_BITS(SPI_USER2_REG(spi_num), SPI_USR_COMMAND_BITLEN, pInData->cmdBitLen - 1,
+                SPI_USR_COMMAND_BITLEN_S);
+        // Enable command
+        SET_PERI_REG_MASK(SPI_USER_REG(spi_num), SPI_USR_COMMAND);
+        // Load command,bit15-0 is cmd value.
+        SET_PERI_REG_BITS(SPI_USER2_REG(spi_num), SPI_USR_COMMAND_VALUE, pInData->cmd, SPI_USR_COMMAND_VALUE_S);
+    } else {
+        CLEAR_PERI_REG_MASK(SPI_USER_REG(spi_num), SPI_USR_COMMAND);
+        SET_PERI_REG_BITS(SPI_USER2_REG(spi_num), SPI_USR_COMMAND_BITLEN, 0, SPI_USR_COMMAND_BITLEN_S);
+    }
+    // Set Address by user.
+    if (pInData->addrBitLen != 0) {
+        SET_PERI_REG_BITS(SPI_USER1_REG(spi_num), SPI_USR_ADDR_BITLEN, (pInData->addrBitLen - 1), SPI_USR_ADDR_BITLEN_S);
+        // Enable address
+        SET_PERI_REG_MASK(SPI_USER_REG(spi_num), SPI_USR_ADDR);
+        // Set address
+        WRITE_PERI_REG(SPI_ADDR_REG(spi_num), *pInData->addr);
+    } else {
+        CLEAR_PERI_REG_MASK(SPI_USER_REG(spi_num), SPI_USR_ADDR);
+        SET_PERI_REG_BITS(SPI_USER1_REG(spi_num), SPI_USR_ADDR_BITLEN, 0, SPI_USR_ADDR_BITLEN_S);
+    }
+    // Set data by user.
+    uint32_t* p_tx_val = pInData->txData;
```

Comment:
`uint32_t* p_tx_val` -> `uint32_t *p_tx_val`

Reply:
fix all * position issue in this file